### ISSUE TITLE
[rule-tester] support multipass fixes

### ISSUE THREAD

**Issue by bradzacher** [typescript-eslint/typescript-eslint#8554]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

https://github.com/eslint/eslint/issues/18007 was rejected as a breaking change enhancement.

Given this impacts us directly (eg https://github.com/typescript-eslint/typescript-eslint/issues/8251) we should build this into our fork.

I propose the following changes:
1) if the test case has `output` and the `output` value is a string and the test case requires more than 1 pass to fix then the test case fails with a message like "test requires multiple fixes to resolve due to overlapping fix ranges - please use the array form of `output` to declare all of the expected fix passes".
2) accept an array for `output` which declares the expected fix passes for a test case.
    - An array of size 0 is disallowed.
    - An array of size 1 is equivalent to the string form.
    - An array of size `n` asserts that there are exactly `n` fix passes
    - The rule tester asserts that after each  fix pass the output matches the `n`th element of the array.

**Comment by bradzacher** [comment]
For me it's a debugging thing.
I remember we had some bugs with one of our rules (array-type) around how multipass fixes worked and we had to setup separate tests to help us debug and validate it.

It's about being explicit and acknowledging that the user will have two fixes applied because your rule defines overlapping ranges. This may be an important thing for perf reasons.

For example imagine switching out `Array<Array<T>>` for `T[][]`. One fix solution would be:
```ts
// fix one
  Array<Array<T>>
//^^^^^^^^^^^^^^^ replace with Array<T>[]

// fix two
  Array<Array<T>>
//      ^^^^^^^^ replace with T[]
```

But that solution is heavy-handed and has overlapping fix ranges so will always require two fix passes.

One might not consider this problem if they don't think about the fact that the test has overlapping fixes.

**Comment by bradzacher** [comment]
This will landed in our v8 release. Please use the `rc-v8` tag to try it out now!
